src/Plugin/AddPathPlugin.php

```diff
@@ -32,17 +41,39 @@ public function __construct(UriInterface $uri)
         }
 
         $this->uri = $uri;
+
+        $resolver = new OptionsResolver();
+        $this->configureOptions($resolver);
+        $options = $resolver->resolve($config);
+
+        $this->smart = $options['smart'];
     }
 
     /**
      * {@inheritdoc}
      */
     public function handleRequest(RequestInterface $request, callable $next, callable $first)
     {
-        $request = $request->withUri($request->getUri()
-            ->withPath($this->uri->getPath().$request->getUri()->getPath())
-        );
+        $prepend = $this->uri->getPath();
+        $path = $request->getUri()->getPath();
+
+        if (!$this->smart || substr($path, 0, strlen($prepend)) !== $prepend) {
```

Comment:
But maybe I'm wrong with the client design and the API path (/api/v3) should be hardcoded in the abstract api class and so, I just need to use the `AddHostPlugin` as the user will only configure it's gitlab instance domain.